sdks/java/io/couchbase/src/main/java/org/apache/beam/sdk/io/couchbase/CouchbaseIO.java

```diff
@@ -0,0 +1,359 @@
+      this.spec = spec;
+      this.cluster = cluster;
+      this.bucket = bucket;
+      this.lowerBound = lb;
+      this.upperBound = ub;
+    }
+
+    private Bucket getBucket() {
+      if (cluster == null) {
+        DefaultCouchbaseEnvironment.Builder builder = DefaultCouchbaseEnvironment.builder();
+        if (spec.httpPort() != null) {
+          builder.bootstrapHttpDirectPort(spec.httpPort());
+        }
+        if (spec.carrierPort() != null) {
+          builder.bootstrapCarrierDirectPort(spec.carrierPort());
+        }
+        cluster = CouchbaseCluster.create(builder.build(), spec.hosts());
+      }
+      if (bucket == null) {
+        // For Couchbase Server, in the previous version than 5.0, the passwordless bucket can be
+        // supported.
+        // But after version 5.0, the newly created user should have a username equal to bucket name
+        // and a password.
+        // For more information, please go to
+        // https://docs.couchbase.com/java-sdk/2.7/sdk-authentication-overview.html#legacy-connection-code
+        bucket =
+            spec.password() == null
+                ? cluster.openBucket(spec.bucket())
+                : cluster.openBucket(spec.bucket(), spec.password());
+      }
+      return bucket;
+    }
+
+    @Override
+    public Coder<JsonDocument> getOutputCoder() {
+      return SerializableCoder.of(JsonDocument.class);
+    }
+
+    @Override
+    public List<? extends BoundedSource<JsonDocument>> split(
+        long desiredBundleSize, PipelineOptions options) {
+      // If the desiredBundleSize equals to 0, it means that there will be only one bundle of data
+      // to be read.
+      int totalBundle =
+          desiredBundleSize == 0 ? 1 : (int) Math.ceil((double) itemCount / desiredBundleSize);
+      List<CouchbaseSource> sources = new ArrayList<>(totalBundle);
+      for (int i = 0, offset = 0; i < totalBundle; i++) {
+        int lowerBound = offset;
+        int upperBound = offset += (int) desiredBundleSize;
+        if (i == totalBundle - 1) {
+          upperBound = itemCount;
+        }
+        sources.add(new CouchbaseSource(spec, cluster, bucket, lowerBound, upperBound));
+      }
+      LOG.debug(String.format("The original source is split to %d sources.", sources.size()));
+      return sources;
+    }
+
+    /**
+     * Attention: The idea is to divide the {@link BoundedSource} by the number of documents. So we
+     * do not estimate the total data size in bytes. In fact, we just fetch the total number of keys
+     * to the target bucket.
+     *
+     * @param options Pipeline options
+     * @return the number of keys to the bucket
+     * @throws Exception throw {@link CouchbaseIOException} when querying Couchbase
+     */
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+      String query = String.format("SELECT RAW COUNT(META().id) FROM `%s`", getBucket().name());
```

Comment:
`if (estimatedSizeBytes == null) ...` Estimating the size of the collection could be a heavy operation so this will save you extra calls.

Reply:
I don't think we need to use `BoundedSource` API since there is no benefit of this here. It could be implemented by using just ParDo. I'll elaborate my point in top-level comment of this review.